src/goto-programs/restrict_function_pointers.cpp

```diff
@@ -0,0 +1,415 @@
+    handler);
+}
+} // namespace
+
+void restrict_function_pointers(
+  goto_modelt &goto_model,
+  const function_pointer_restrictionst &restrictions)
+{
+  typecheck_function_pointer_restrictions(goto_model, restrictions);
+
+  for(auto &goto_function : goto_model.goto_functions.function_map)
+  {
+    auto &goto_function_body = goto_function.second.body;
+    // for each function call, we check if it is using a symbol we have
+    // restrictions for, and if so branch on its value and insert concrete calls
+    // to the restriction functions
+    for_each_function_call(
+      goto_function.second, [&](const goto_programt::targett location) {
+        // Check if this is calling a function pointer, and if so if it is one
+        // we have a restriction for
+        const auto &original_function_call = location->get_function_call();
+        if(can_cast_expr<dereference_exprt>(original_function_call.function()))
+        {
+          // because we simplify before this stage a dereference can only
+          // dereference a symbol expression
+          auto const &called_function_pointer =
+            to_dereference_expr(original_function_call.function()).pointer();
+          PRECONDITION(can_cast_expr<symbol_exprt>(called_function_pointer));
+          auto const &pointer_symbol = to_symbol_expr(called_function_pointer);
+          auto const restriction_iterator =
+            restrictions.restrictions.find(pointer_symbol.get_identifier());
+          if(restriction_iterator != restrictions.restrictions.end())
+          {
+            auto const &restriction = *restriction_iterator;
+            // if we can, we will replace uses of it by the
+            // this is intentionally a copy because we're just about to change
+            // the instruction this iterator points to
+            auto const original_function_call_instruction = *location;
+            *location = goto_programt::make_assertion(
+              false_exprt{},
+              make_function_pointer_restriction_assertion_source_location(
+                original_function_call_instruction.source_location,
+                restriction));
+            auto const assume_false_location = goto_function_body.insert_after(
+              location,
+              goto_programt::make_assumption(
+                false_exprt{},
+                original_function_call_instruction.source_location));
+            // this is mutable because we'll update this at the end of each
+            // loop iteration to always point at the start of the branch
+            // we created
+            auto else_location = location;
+            auto const end_if_location = goto_function_body.insert_after(
+              assume_false_location, goto_programt::make_skip());
+            for(auto const &restriction_target : restriction.second)
+            {
+              auto new_instruction = original_function_call_instruction;
```

Comment:
:bulb: This definitively should be pulled out into a separate function, independently unit tested to get an idea of what goto code it produces.

Reply:
Put this into a separate function. I think it is not necessary to add unit tests as this functionality is very well tested by the regression tests already.